public/js/components/specs/app.spec.js

```diff
@@ -0,0 +1,243 @@
+  test("component can mount", () => {
+    const wrapper = shallowMountApp();
+
+    expect(wrapper.isVueInstance()).toBeTruthy();
+  });
+
+  test("if child components mount", () => {
+    const wrapper = shallowMountApp();
+
+    expect(wrapper.find({ name: "dashboard-settings" }).exists()).toBe(true);
+    expect(wrapper.find({ name: "base-card" }).exists()).toBe(true);
+  });
+
+  test("check before component is created that the default data is clean", () => {
+    const { vm } = shallowMount(App);
+
+    expect(vm.dashboard.blockSize).toEqual([]);
+    expect(vm.dashboard.tiles).toEqual([]);
+    expect(vm.messages).toEqual([]);
+    expect(vm.deviceList).toEqual([]);
+    expect(vm.simulating).toEqual(SIMULATING);
+  });
+
+  test("computed value returned from dashStyle computed method", () => {
+    const { vm } = shallowMountApp();
+
+    vm.dashboard.bgImageRepeat = mockDashboardData.dashboard.bgImageRepeat;
+
+    expect(vm.dashStyle).toEqual({
+      backgroundColor: vm.dashboard.bgColor,
+      backgroundImage: vm.dashboard.bgImageUrl,
+      backgroundRepeat: "repeat"
+    });
+
+    vm.dashboard.bgImageRepeat = false;
+
+    expect(vm.dashStyle).toEqual({
+      backgroundColor: vm.dashboard.bgColor,
+      backgroundImage: vm.dashboard.bgImageUrl,
+      backgroundRepeat: "no-repeat"
+    });
+  });
+
+  test("h1 heading style based upon headingStyle computed method", () => {
+    const { vm } = shallowMountApp();
+
+    expect(vm.headingStyle).toEqual({ color: "#000" });
+  });
+
+  test("don't display dashboard-settings based upon showSettings computed method", () => {
+    const wrapper = shallowMountApp();
+
+    expect(wrapper.vm.showSettings).toEqual(true);
+
+    wrapper.vm.dashboard.editMode = "locked";
+
+    expect(wrapper.vm.showSettings).toEqual(false);
+    expect(wrapper.find({ name: "dashboard-settings" }).exists()).toBe(false);
+  });
+
+  test("compute the appTitle with or without an emoji using the TITLE_EMOJI_REGEX constant", () => {
+    const { vm } = shallowMountApp();
+    const title = TITLE_EMOJI_REGEX.exec(mockDashboardData.dashboard.title);
+
+    title[1] = "\u2700";
+    title[2] = "IoT Dashboard";
+
+    expect(vm.appTitle).toEqual(
+      `<span class="hemoji">${title[1]}</span> ${title[2]}`
+    );
+  });
+
+  test("onSaveSettings method", () => {
+    const { vm } = shallowMountApp();
+
+    expect(vm.dashboard).toEqual(
+      expect.objectContaining({
+        bgColor: expect.any(String),
+        bgImageRepeat: expect.any(Boolean),
+        bgImageUrl: expect.any(String),
+        blockSize: expect.any(Array),
+        editMode: expect.any(String),
+        tiles: expect.any(Array),
+        title: expect.any(String)
+      })
+    );
+
+    vm.onSaveSettings(event => {
+      event = {
+        bgColor: "#fff",
+        bgImageRepeat: true,
+        bgImageUrl: "",
+        title: "\u2700 IoT Dashboard"
+      };
+      expect(vm.saveDashboard).toHaveBeenCalled();
```

Comment:
Good call, this is a really good point. I vaguely remember this would fail if it was outside of the vm.onSaveSettings (and on vm.onTileChange on line 162). I'll go ahead and remove this.